src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs

```diff
@@ -1,65 +1,128 @@
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using System.Reflection;
 using System.Runtime.InteropServices;
 using Microsoft.DotNet.Cli.Compiler.Common;
 using Microsoft.DotNet.Cli.Utils;
 using Microsoft.DotNet.ProjectModel;
 using Microsoft.DotNet.ProjectModel.Compilation;
 using Microsoft.DotNet.ProjectModel.Files;
+using Microsoft.DotNet.ProjectModel.Graph;
 using Microsoft.Extensions.DependencyModel;
 using NuGet.Frameworks;
 
 namespace Orchard.Environment.Extensions.Compilers
 {
-    public interface ICompiler
+    public class CSharpExtensionCompiler
     {
-        bool Compile(ProjectContext context, string config, string buildBasePath);
-    }
+        public const string RefsDirectoryName = "refs";
+        private static readonly Object _syncLock = new Object();
+        private static readonly ConcurrentDictionary<string, object> _compilationlocks = new ConcurrentDictionary<string, object>();
+        private static readonly ConcurrentDictionary<string, bool> _ambientLibraries = new ConcurrentDictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+        private static readonly ConcurrentDictionary<string, bool> _compiledLibraries = new ConcurrentDictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+        private static readonly Lazy<Assembly> _entryAssembly = new Lazy<Assembly>(Assembly.GetEntryAssembly);
+
+        public CSharpExtensionCompiler ()
+        {
+            Diagnostics = new List<string>();
+        }
 
-    public class CSharpExtensionCompiler : ICompiler
-    {
-        public bool Compile(ProjectContext context, string config, string buildBasePath)
+        public static Assembly EntryAssembly => _entryAssembly.Value;
+        public IList<string> Diagnostics { get; private set; }
+
+        public bool Compile(ProjectContext context, string config, string probingFolderPath)
+        {
+            var compilationlock = _compilationlocks.GetOrAdd(context.RootProject.Identity.Name, id => new object());
+
+            lock (compilationlock)
+            {
+                return CompileInternal(context, config, probingFolderPath);
+            }
+        }
+
+        internal bool CompileInternal(ProjectContext context, string config, string probingFolderPath)
         {
-            // Set up Output Paths
-            var outputPaths = context.GetOutputPaths(config, buildBasePath);
+            // Mark ambient libraries as compiled
+            if (_ambientLibraries.IsEmpty)
+            {
+                var libraries = DependencyContext.Default.CompileLibraries
+                    .Where(x => x.Type == LibraryType.Project.ToString().ToLowerInvariant());
```

Comment:
string.Equals(x.Type, LibraryType.Project.ToString(), StringComparision.Original  <- You know wha I mean (I forgot the last bit))